### PR TITLE
Fix object detail browsing

### LINKED ISSUE #5837: Rethink UX of viewing a single record's details
Now that we have the ability to page through single records (object details), I started thinking about how our current implementation of the detail or single-record view is a little off. 

For queries that return a single record, I think we should keep doing what we're doing. But the other use case is looking at a segment or subset of not-necessarily-sequential records (in terms of their ID numbers that is), and paging through each record one at a time. This second use case doesn't really work well right now, for two reasons:
1. When clicking on a single record's ID to "view details," we run a new query (devoid of any notion of the context of the originating segment or filtered table) instead of showing a "zoomed in" view of that record. This means that if I then want to go to the "next" record, I'll be paging through _all_ the records; not just the records included in the list I had started from.
2. The current implementation of the next/previous record arrows just simple increments or decrements the ID in the filter, which of course doesn't work if the IDs aren't simple sequential integers, and also doesn't work if there are "missing" numbers (you'll get a "no results" message). 

Both of these could be solved by implementing a kind of single record zoom view, instead of running a totally new query. Sort of like opening a single row in a modal, with the ability to close it to go back to the list without re-running the list query, and allowing the user to page through the current list while in the single-record view.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by kulyk**
Resolves #5837. People can click PK and FK cells in questions' result tables and end up on the object detail view — a separate visualization aimed to display individual records. Object detail views also have "previous" and "next" buttons to go through records one by one.

### Background & Limitations

The way it works is pretty straightforward. Let's say I have a question selecting everything from the sample orders table. Once I click an "ID" column cell with the value "5", Metabase will apply a filter to the question to only show records with `ID === 5`. And when I click "next" or "previous", it will increment or decrement the ID value in the filter. There are a few disadvantages with this approach:

1. It won't preserve the order of the records as we're just doing `ID++` and `ID--`, though a question can be filtered/sorted
2. It doesn't work at all for non-numeric IDs (say UUIDs)
3. Every click on the "previous" and "next" button results in a new query, though we have all the data needed in the original question

### How it works now

**Primary keys**

1. This PR adds [`getPKRowIndexMap` selector](https://github.com/metabase/metabase/blob/7afc7aa0b6c30d202429bcc5d9a2b89aef840cb7/frontend/src/metabase/query_builder/selectors.js#L101). It's computed from the query results and it basically allows to quickly get record's row index knowing its PK value.
2. Once a user clicks a PK column cell value, the drill handler dispatches an action with this ID and we put the ID in QB's redux state
3. Having the clicked ID value and a map from `getPKRowIndexMap`, we can get the row index for the clicked row
4. We put the object ID in the URL (e.g. `/question/1-orders/5` where `5` is the order's ID value). As `getPKRowIndexMap` is computed out of query results, it should always open the correct order as we don't rely on the row index here.

**Foreign keys**

Foreign keys drills used to work in a similar way to PKs. For FKs, Metabase would just throw away the original question and replace it with the linked table, and then add a filter for the linked table's ID field. This is how it works now:

1. User clicks on the FK cell (let's say a "Product ID" in Order's table)
2. Click handler returns a URL like `/question?objectId=5#serializedQueryToThePeopleTable`, where `5` is a User ID
3. Navigating to a new URL inside the QB triggers the `initializeQB` action
4. `initializeQB` parses the serialized card and the object ID from query parameters
5. Object ID from the query params is stored in QB redux
6. Once the query to People table execs, we can compute `getPKRowIndexMap` and get the desired row index knowing the object ID
7. Ta-daaa

**Composite primary keys**

This is a use case that can use some extra effort. For simplicity, for composite PKs we'd still use filters and it should work the same way as when it was fixed in #18016:

1. Let's say we have two columns (e.g. `REPO_ID` and `PR_ID` as we have some GitHub PRs tables)
2. We click `REPO_ID` (say `5`)
3. We apply a filter like `REPO_ID = 5`
4. We get a table listing all PR's with `REPO_ID = 5`
5. We click any `PR_ID`
6. Once we have "=" filters for every PK column in a table, we finally land in the object detail view as it's guaranteed to have only one record
7. We hide the "previous" and "next" buttons as there's no clear way to get the prev/next row. We display them on `master` in this case currently, but clicking it will actually show a "No results" page
8. We don't include object ID in the URL for now as it'd require another URL scheme (something like `/question/1?repo_id=5&pr_id=10`

### CODE REVIEW

File: frontend/test/metabase/scenarios/dashboard/dashboard-drill.cy.spec.js [hunk 1]
```diff
@@ -599,7 +599,7 @@ describe("scenarios > dashboard > dashboard drill", () => {
 
   it('should drill-through on PK/FK to the "object detail" when filtered by explicit joined column (metabase#15331)', () => {
     cy.server();
-    cy.route("POST", "/api/dataset").as("dataset");
+    cy.route("POST", "/api/card/*/query").as("cardQuery");
```
Comment (kulyk) [inline frontend/test/metabase/scenarios/dashboard/dashboard-drill.cy.spec.js:602]: Not using `/api/dataset` when browsing object details anymore because the filters approach is replaced 🎉

File: frontend/test/metabase/scenarios/question/new.cy.spec.js [hunk 2]
```diff
@@ -299,18 +299,11 @@ describe("scenarios > question > new", () => {
     });
 
     it("should correctly choose between 'Object Detail' and 'Table (metabase#13717)", () => {
-      // set ID to `No semantic type`
-      cy.request("PUT", `/api/field/${ORDERS.ID}`, {
-        semantic_type: null,
-      });
-      // set Quantity to `Entity Key`
```
Comment (kulyk) [inline frontend/test/metabase/scenarios/question/new.cy.spec.js:306]: This is a test for composite PK scenario (> 1 PK columns), but making the "Quantity" a PK and removing the PK type for "ID" doesn't make sense in this case

File: frontend/src/metabase/query_builder/actions.js [hunk 3]
```diff
@@ -154,6 +158,22 @@ export const popState = createThunkAction(
...
     dispatch(cancelQuery());
+
+    const zoomedObjectId = getZoomedObjectId(getState());
+    if (zoomedObjectId) {
+      const { locationBeforeTransitions } = getState().routing;
+
+      const previouslyZoomedObjectId =
+        locationBeforeTransitions?.state?.objectId;
+
+      if (previouslyZoomedObjectId) {
```
Comment (alxnddr) [inline frontend/src/metabase/query_builder/actions.js:169]: For some reason, if you zoom on object 1, go to object 2, go back to the previous, and then go back to question — it keeps `previouslyZoomedObjectId` as object 1 and you can't return to your question. Do you know what can be the reason?

**Review by alxnddr — COMMENTED** [review]
Fantastic improvement in UX and correctness 👏 

I found a couple of issues that it would be great to fix:

1) Menu item width
Screenshot 2022-02-09 at 21 19 51

2) Unable to go back to an ad-hoc query:
- Zoom
- Go to the next object
- Try go back to a question 
**It is not possible to return back to a question**

3) Ad-hoc query keeps object id param without zoom
- Zoom
- Go to the next object twice
- Go back 
**It instantly shows the question instead of the previous object. The URL contains object id so that on a page refresh it opens the object view**

**Review by mazameli — APPROVED** [review]
This is so fast now, and it's great that the next and prev buttons work corretly.

**Comment by kulyk** [comment]
Great catches @alxnddr, thank you! 🙏 There were a few missed edge cases in the `popState` handler, but they should be fixed now. The "view details" button width should now also be OK

**Review by alxnddr — APPROVED** [review]
👍

File: frontend/test/metabase/scenarios/dashboard/dashboard-drill.cy.spec.js (hunk 1 above)
Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/dashboard/dashboard-drill.cy.spec.js:602]: @kulyk Note for the future - when you encounter `cy.server()` and `cy.route()` and you need to change something, please also change it to `cy.intercept()`.

**Comment by diogormendes** [comment]
@flamber , should we rever this in favor of fixing #21756 ?

**Comment by flamber** [comment]
I have no idea, it's just a simple revert - not with all the other changes that has happened. And it would reopen the old issue, which is also problematic.
